### PR TITLE
Notify checkpoint listeners after init and replay

### LINKED ISSUE #9916: Notify CheckpointListeners after init and replay
**Description**

InterpartitionCommandSender and Reciever keeps track of the latest checkpointId via CheckpointListeners. We must ensure that they have the correct valued even after failovers or restart. Currently CheckpointListeners are notified only when checkpoint record is processed. In addition, they should be notified after the processor initializes it's state, when a new listener is registerd, and CREATED record is replayed.

Ref:- https://github.com/camunda/zeebe/pull/9899#discussion_r932089371

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

The listeners must be upddated with the current checkpointId after init and replay. Otherwise if there are no new checkpoint created, after a restart or failover the listeners cannot know about the latest checkpoint. It is important for `InterPartitionCommandSencer` and `InterPartitionCommandReceiver` to keep track of the latest checkpointId to guarantee that the checkpoints are consistent.

## Related issues

closes #9916 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: backup/src/main/java/io/camunda/zeebe/backup/processing/CheckpointRecordsProcessor.java [hunk 1]
```diff
@@ -80,4 +95,13 @@ public ProcessingResult onProcessingError(
     // from making progress.
     throw new RuntimeException(processingException);
   }
+
+  /**
+   * Registers a listener.
+   *
+   * @param checkpointListener
+   */
+  public void addCheckpointListener(final CheckpointListener checkpointListener) {
```
Comment (oleschoenburg) [inline backup/src/main/java/io/camunda/zeebe/backup/processing/CheckpointRecordsProcessor.java:104]: :x: If I remember correctly (and this matches with the original issue description), we need to call the listener here directly with the current checkpoint id.

File: backup/src/main/java/io/camunda/zeebe/backup/api/CheckpointListener.java [hunk 2]
```diff
@@ -13,8 +13,10 @@ public interface CheckpointListener {
   /**
    * Called when ever a new checkpoint is created.
    *
-   * <p>Will be called immediately after CHECKPOINT:CREATE record is processed if it results in a
-   * new checkpoint.
+   * <p>This will be called
+   * <li>When the processor is initialized with the latest checkpoint
+   * <li>When CHECKPOINT:CREATE record is processed and if it results in a new checkpoint.
+   * <li>When CHECKPOINT:CREATED record is replayed
```
Comment (oleschoenburg) [inline backup/src/main/java/io/camunda/zeebe/backup/api/CheckpointListener.java:19]: :wrench: we should add that it is called on registering

**Review by lenaschoenburg — CHANGES_REQUESTED** [review]
Looks good, except that I think that we need to call the listener on register.

File: backup/src/main/java/io/camunda/zeebe/backup/processing/CheckpointRecordsProcessor.java (hunk 1 above)
Comment (deepthidevaki) [reply backup/src/main/java/io/camunda/zeebe/backup/processing/CheckpointRecordsProcessor.java:104]: :+1: I had added it and then removed it :) But, you are right. It is safer to notify when registering to ensure that it doesn't miss an update. I will update it.

**Review by lenaschoenburg — APPROVED** [review]
Awesome :rocket:

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**